src/piece_picker.cpp

```diff
@@ -2149,6 +2153,7 @@ namespace {
 						pc.inc_stats_counter(counters::piece_picker_reverse_rare_loops);
 
 						if (!is_piece_free(m_pieces[p], pieces)) continue;
+						if ((options & sequential) && m_pieces[p] >= m_cursor && m_pieces[p] < m_reverse_cursor) continue;
```

Comment:
I don't understand why this condition is here. We would only get here if all pieces in the sequential range had been picked already, in which case they wouldn't be considered free.

Do you have a test case that exercises this condition?

If there is a good reason for this, I think there should be a comment in the code explaining it.

Reply:
If I recall a piece not being free refers to a piece we have, not a piece we intend to be picked. If this check weren't here then duplicate blocks ("pieces") would be picked. I put this predicate right after the is_piece_free predicate to render it self-explanatory.

The three subtests for sequential range downloads exercise this scenario. I just double checked them and this is necessary.

Reply:
you're right that [is_piece_free()](https://github.com/arvidn/libtorrent/blob/RC_2_0/src/piece_picker.cpp#L2585) is still true after we have picked all blocks in it.

So, the intention with this is to not waste time considering pieces that should already have been considered in the sequential-loop above. Is that right?

I don't think we get to this case in the unit test though, since we exhaust all sequential pieces in the first pass, set `m_cursor` and `m_reverse_cursor` to an invalid range, which means the piece will never be within the range here.

I propose a new test case, where we request sequential, but the peer we request from is missing one of the sequential pieces. After all sequential pieces have been picked, we should fall through into here and pick rarest first and actually hit this case.

Reply:
Do the test actually take the `continue` branch of the if-statement?

Reply:
The intention with this is not merely not to waste time. It is actually necessary for correctness; otherwise duplicate pieces may be picked if the `continue` branch is not taken, which the subtests exercise to ensure correctness. I suggest you look closely at the three subtests to convince yourself of this.

You forget that the cursors are never mutated during piece picking. Only once we have pieces are they possibly mutated.

Reply:
would you mind adding a comment to the code, explaining this?